File: supabase/functions/_shared/uk_population.ts
import Papa from "https://esm.sh/papaparse@5.4.1?target=deno";
import * as XLSX from "https://esm.sh/xlsx@0.18.5?target=deno";

export type CatalogueEntry = {
  geography_level: string;
  geography_vintage: string;
  coverage: string;
  nomis_dataset_api_ref: string | null;
  nomis_dataset_keyfamily_id: string | null;
  nomis_geography_type_code: string | null;
  typical_update_cycle: string | null;
  latest_reference_period_on_nomis: string | null;
  next_release_note: string | null;
  notes: string | null;
  primary_source_org: string | null;
  primary_source_dataset_page: string | null;
  primary_source_download_csv: string | null;
  primary_source_download_xlsx: string | null;
  primary_source_api_example: string | null;
  primary_source_years_available: string | null;
  primary_source_update_frequency: string | null;
  primary_source_notes: string | null;
};

export type IngestOverride = {
  dataset_id: string;
  source_url?: string | null;
  format?: string | null;
  geo_code_column?: string | null;
  time_column?: string | null;
  value_column?: string | null;
  measure_column?: string | null;
  jsonstat_geo_dimension?: string | null;
  jsonstat_time_dimension?: string | null;
  jsonstat_value_dimension?: string | null;
};

const normalizeValue = (value: unknown): string | null => {
  if (value === null || value === undefined) {
    return null;
  }
  const text = String(value).trim();
  return text.length > 0 ? text : null;
};

export const normalizeCatalogueEntry = (row: Record<string, unknown>): CatalogueEntry => {
  const geography_level = normalizeValue(row.geography_level);
  const geography_vintage = normalizeValue(row.geography_vintage);
  const coverage = normalizeValue(row.coverage);
  if (!geography_level || !geography_vintage || !coverage) {
    throw new Error("Catalogue row missing required fields.");
  }
  return {
    geography_level,
    geography_vintage,
    coverage,
    nomis_dataset_api_ref: normalizeValue(row.nomis_dataset_api_ref),
    nomis_dataset_keyfamily_id: normalizeValue(row.nomis_dataset_keyfamily_id),
    nomis_geography_type_code: normalizeValue(row.nomis_geography_type_code),
    typical_update_cycle: normalizeValue(row.typical_update_cycle),
    latest_reference_period_on_nomis: normalizeValue(row.latest_reference_period_on_nomis),
    next_release_note: normalizeValue(row.next_release_note),
    notes: normalizeValue(row.notes),
    primary_source_org: normalizeValue(row.primary_source_org),
    primary_source_dataset_page: normalizeValue(row.primary_source_dataset_page),
    primary_source_download_csv: normalizeValue(row.primary_source_download_csv),
    primary_source_download_xlsx: normalizeValue(row.primary_source_download_xlsx),
    primary_source_api_example: normalizeValue(row.primary_source_api_example),
    primary_source_years_available: normalizeValue(row.primary_source_years_available),
    primary_source_update_frequency: normalizeValue(row.primary_source_update_frequency),
    primary_source_notes: normalizeValue(row.primary_source_notes),
  };
};

export const parseCatalogueCsv = (csvText: string): CatalogueEntry[] => {
  const parsed = Papa.parse<Record<string, string>>(csvText, {
    header: true,
    skipEmptyLines: true,
  });
  if (parsed.errors?.length) {
    const message = parsed.errors.map((err) => err.message).join("; ");
    throw new Error(`Catalogue CSV parse failed: ${message}`);
  }
  return parsed.data.map((row) => normalizeCatalogueEntry(row));
};

export const assignSourcePrefix = (entry: CatalogueEntry): "nomis" | "nrs" | "nisra" => {
  if (entry.nomis_dataset_api_ref || entry.nomis_dataset_keyfamily_id || entry.nomis_geography_type_code) {
    return "nomis";
  }
  const coverage = entry.coverage.toLowerCase();
  if (coverage.includes("scotland")) {
    return "nrs";
  }
  if (coverage.includes("northern ireland")) {
    return "nisra";
  }
  if (coverage.includes("england") || coverage.includes("wales")) {
    return "nomis";
  }
  return "nomis";
};

export const slugify = (value: string): string => {
  return value.trim().toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_+|_+$/g, "");
};

export const buildDatasetId = (prefix: string, entry: CatalogueEntry): string => {
  const parts = [
    prefix,
    slugify(entry.geography_level),
    slugify(entry.geography_vintage),
    slugify(entry.coverage),
  ].filter((part) => part.length > 0);
  return parts.join("_");
};

export const pickSourceUrl = (entry: CatalogueEntry): string | null => {
  return entry.primary_source_download_csv ??
    entry.primary_source_download_xlsx ??
    entry.primary_source_api_example ??
    null;
};

export const inferFormat = (url: string): string => {
  const lowered = url.toLowerCase();
  if (lowered.endsWith(".csv")) {
    return "csv";
  }
  if (lowered.endsWith(".xlsx") || lowered.endsWith(".xls")) {
    return "xlsx";
  }
  if (lowered.includes(".json")) {
    return "jsonstat";
  }
  return "csv";
};

export const inferColumn = (headers: string[], candidates: string[]): string | null => {
  const lowerHeaders = new Map(headers.map((header) => [header.toLowerCase(), header]));
  for (const candidate of candidates) {
    const match = lowerHeaders.get(candidate.toLowerCase());
    if (match) {
      return match;
    }
  }
  for (const header of headers) {
    const lower = header.toLowerCase();
    for (const candidate of candidates) {
      if (lower.includes(candidate.toLowerCase())) {
        return header;
      }
    }
  }
  return null;
};

export const parseReferenceDateFlexible = (value: unknown): string => {
  const text = String(value ?? "").trim();
  if (/^\d{4}$/.test(text)) {
    return `${text}-06-30`;
  }
  if (/^\d{4}-\d{2}-\d{2}$/.test(text)) {
    return text;
  }
  const match = text.match(/(19|20)\d{2}/);
  if (match) {
    return `${match[0]}-06-30`;
  }
  throw new Error(`Unsupported time value: ${text}`);
};

export const fetchCsvRows = async (url: string): Promise<Record<string, unknown>[]> => {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`CSV request failed (${response.status}): ${url}`);
  }
  const text = await response.text();
  const parsed = Papa.parse<Record<string, string>>(text, {
    header: true,
    skipEmptyLines: true,
  });
  if (parsed.errors?.length) {
    const message = parsed.errors.map((err) => err.message).join("; ");
    throw new Error(`CSV parse failed: ${message}`);
  }
  return parsed.data;
};

export const fetchXlsxRows = async (url: string): Promise<Record<string, unknown>[]> => {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`XLSX request failed (${response.status}): ${url}`);
  }
  const buffer = await response.arrayBuffer();
  const workbook = XLSX.read(buffer, { type: "array" });
  const sheetName = workbook.SheetNames[0];
  if (!sheetName) {
    return [];
  }
  const sheet = workbook.Sheets[sheetName];
  return XLSX.utils.sheet_to_json(sheet, { defval: null });
};

const buildCategory = (index: Record<string, number> | string[]) => {
  if (Array.isArray(index)) {
    const map = new Map<string, number>();
    index.forEach((key, idx) => map.set(key, idx));
    return { keys: index, indexOf: (key: string) => map.get(key) ?? -1 };
  }
  const entries = Object.entries(index).sort((a, b) => a[1] - b[1]);
  const keys = entries.map(([key]) => key);
  const map = new Map(entries);
  return { keys, indexOf: (key: string) => map.get(key) ?? -1 };
};

export const fetchJsonStatRows = async (
  url: string,
  geoDimension: string,
  timeDimension: string,
  valueDimension?: string | null,
): Promise<Record<string, unknown>[]> => {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`JSON-stat request failed (${response.status}): ${url}`);
  }
  const payload = await response.json();
  const dataset = payload?.dataset ?? payload;
  const dimension = dataset?.dimension ?? {};
  const ids = dataset?.id ?? [];
  const sizes = dataset?.size ?? [];
  const values = dataset?.value ?? [];

  const idToPos = new Map<string, number>();
  ids.forEach((id: string, idx: number) => idToPos.set(id, idx));
  const geoPos = idToPos.get(geoDimension);
  const timePos = idToPos.get(timeDimension);
  const valuePos = valueDimension ? idToPos.get(valueDimension) : undefined;

  if (geoPos === undefined || timePos === undefined) {
    throw new Error("JSON-stat dimensions not found.");
  }

  const geoCats = buildCategory(dimension[geoDimension]?.category?.index ?? {});
  const timeCats = buildCategory(dimension[timeDimension]?.category?.index ?? {});
  const valueCats = valueDimension
    ? buildCategory(dimension[valueDimension]?.category?.index ?? {})
    : { keys: [null], indexOf: () => 0 };

  const rows: Record<string, unknown>[] = [];
  for (const geo of geoCats.keys) {
    for (const time of timeCats.keys) {
      for (const valueKey of valueCats.keys) {
        const coords = new Array(ids.length).fill(0);
        coords[geoPos] = geoCats.indexOf(geo);
        coords[timePos] = timeCats.indexOf(time);
        if (valuePos !== undefined && valueDimension) {
          coords[valuePos] = valueCats.indexOf(String(valueKey));
        }
        let index = 0;
        for (let i = 0; i < sizes.length; i += 1) {
          index = index * sizes[i] + coords[i];
        }
        const value = values[index];
        const row: Record<string, unknown> = {
          [geoDimension]: geo,
          [timeDimension]: time,
          value,
        };
        if (valuePos !== undefined && valueDimension) {
          row[valueDimension] = valueKey;
        }
        rows.push(row);
      }
    }
  }
  return rows;
};

export const mapRows = (
  rows: Record<string, unknown>[],
  geoType: string,
  datasetId: string,
  geoCodeColumn: string,
  timeColumn: string,
  valueColumn: string,
  measureColumn?: string | null,
): Record<string, unknown>[] => {
  return rows.map((row) => {
    const referenceDate = parseReferenceDateFlexible(row[timeColumn]);
    const populationValue = Number(row[valueColumn]);
    if (!Number.isFinite(populationValue)) {
      throw new Error(`Invalid population value: ${row[valueColumn]}`);
    }
    return {
      geo_type: geoType.trim().toUpperCase(),
      geo_code: String(row[geoCodeColumn] ?? "").trim(),
      reference_date: referenceDate,
      population_value: Math.trunc(populationValue),
      dataset_id: datasetId,
      measure: measureColumn ? row[measureColumn] ?? null : null,
    };
  });
};

File: src/nomis_api/uk_population.py
from dataclasses import dataclass
from datetime import date
import csv
import io
import json
from pathlib import Path
import re
from typing import Iterable

import requests

from nomis_api.geo import normalize_geo_type
from nomis_api.ingest import parse_reference_date


@dataclass(frozen=True)
class CatalogueEntry:
    geography_level: str
    geography_vintage: str
    coverage: str
    nomis_dataset_api_ref: str | None
    nomis_dataset_keyfamily_id: str | None
    nomis_geography_type_code: str | None
    typical_update_cycle: str | None
    latest_reference_period_on_nomis: str | None
    next_release_note: str | None
    notes: str | None
    primary_source_org: str | None
    primary_source_dataset_page: str | None
    primary_source_download_csv: str | None
    primary_source_download_xlsx: str | None
    primary_source_api_example: str | None
    primary_source_years_available: str | None
    primary_source_update_frequency: str | None
    primary_source_notes: str | None


@dataclass(frozen=True)
class IngestOverride:
    dataset_id: str
    source_url: str | None = None
    format: str | None = None
    geo_code_column: str | None = None
    time_column: str | None = None
    value_column: str | None = None
    measure_column: str | None = None
    jsonstat_geo_dimension: str | None = None
    jsonstat_time_dimension: str | None = None
    jsonstat_value_dimension: str | None = None


def load_catalogue(path: Path) -> list[CatalogueEntry]:
    rows = []
    with path.open(newline="", encoding="utf-8") as handle:
        reader = csv.DictReader(handle)
        for row in reader:
            rows.append(
                CatalogueEntry(
                    geography_level=row["geography_level"],
                    geography_vintage=row["geography_vintage"],
                    coverage=row["coverage"],
                    nomis_dataset_api_ref=row.get("nomis_dataset_api_ref") or None,
                    nomis_dataset_keyfamily_id=row.get("nomis_dataset_keyfamily_id") or None,
                    nomis_geography_type_code=row.get("nomis_geography_type_code") or None,
                    typical_update_cycle=row.get("typical_update_cycle") or None,
                    latest_reference_period_on_nomis=row.get("latest_reference_period_on_nomis") or None,
                    next_release_note=row.get("next_release_note") or None,
                    notes=row.get("notes") or None,
                    primary_source_org=row.get("primary_source_org") or None,
                    primary_source_dataset_page=row.get("primary_source_dataset_page") or None,
                    primary_source_download_csv=row.get("primary_source_download_csv") or None,
                    primary_source_download_xlsx=row.get("primary_source_download_xlsx") or None,
                    primary_source_api_example=row.get("primary_source_api_example") or None,
                    primary_source_years_available=row.get("primary_source_years_available") or None,
                    primary_source_update_frequency=row.get("primary_source_update_frequency") or None,
                    primary_source_notes=row.get("primary_source_notes") or None,
                )
            )
    return rows


def load_ingest_overrides(path: Path) -> dict[str, IngestOverride]:
    payload = json.loads(path.read_text())
    overrides = {}
    for item in payload.get("overrides", []):
        override = IngestOverride(**item)
        overrides[override.dataset_id] = override
    return overrides


def assign_source_prefix(entry: CatalogueEntry) -> str:
    if entry.nomis_dataset_api_ref or entry.nomis_dataset_keyfamily_id or entry.nomis_geography_type_code:
        return "nomis"
    coverage = entry.coverage.lower()
    if "scotland" in coverage:
        return "nrs"
    if "northern ireland" in coverage:
        return "nisra"
    if "england" in coverage or "wales" in coverage:
        return "nomis"
    return "nomis"


def slugify(value: str) -> str:
    value = value.strip().lower()
    value = re.sub(r"[^a-z0-9]+", "_", value)
    return value.strip("_")


def build_dataset_id(prefix: str, entry: CatalogueEntry) -> str:
    parts = [
        prefix,
        slugify(entry.geography_level),
        slugify(entry.geography_vintage),
        slugify(entry.coverage),
    ]
    return "_".join([part for part in parts if part])


def pick_source_url(entry: CatalogueEntry) -> str | None:
    return (
        entry.primary_source_download_csv
        or entry.primary_source_download_xlsx
        or entry.primary_source_api_example
    )


def infer_format(url: str) -> str:
    lowered = url.lower()
    if lowered.endswith(".csv"):
        return "csv"
    if lowered.endswith(".xlsx") or lowered.endswith(".xls"):
        return "xlsx"
    if ".json" in lowered:
        return "jsonstat"
    return "csv"


def fetch_csv_rows(url: str) -> list[dict]:
    response = requests.get(url, timeout=120)
    response.raise_for_status()
    text = response.text
    reader = csv.DictReader(io.StringIO(text))
    return list(reader)


def fetch_xlsx_rows(url: str) -> list[dict]:
    from openpyxl import load_workbook

    response = requests.get(url, timeout=120)
    response.raise_for_status()
    workbook = load_workbook(io.BytesIO(response.content), read_only=True, data_only=True)
    sheet = workbook.active
    rows = list(sheet.iter_rows(values_only=True))
    if not rows:
        return []
    headers = [str(value).strip() if value is not None else "" for value in rows[0]]
    data_rows = []
    for row in rows[1:]:
        mapped = {}
        for idx, header in enumerate(headers):
            if header:
                mapped[header] = row[idx] if idx < len(row) else None
        data_rows.append(mapped)
    return data_rows


def ordered_categories(category_index: dict) -> list[str]:
    return [key for key, _ in sorted(category_index.items(), key=lambda item: item[1])]


def fetch_jsonstat_rows(
    url: str,
    geo_dimension: str,
    time_dimension: str,
    value_dimension: str | None,
) -> list[dict]:
    response = requests.get(url, timeout=120)
    response.raise_for_status()
    payload = response.json()
    dataset = payload.get("dataset", payload)
    dimension = dataset["dimension"]
    ids = dataset["id"]
    sizes = dataset["size"]
    values = dataset["value"]

    id_to_pos = {dimension_id: idx for idx, dimension_id in enumerate(ids)}
    geo_pos = id_to_pos[geo_dimension]
    time_pos = id_to_pos[time_dimension]
    value_pos = id_to_pos[value_dimension] if value_dimension and value_dimension in id_to_pos else None

    geo_cats = ordered_categories(dimension[geo_dimension]["category"]["index"])
    time_cats = ordered_categories(dimension[time_dimension]["category"]["index"])
    value_cats = (
        ordered_categories(dimension[value_dimension]["category"]["index"]) if value_pos is not None else [None]
    )

    rows = []
    for geo in geo_cats:
        for time in time_cats:
            for value_key in value_cats:
                coords = [0] * len(ids)
                coords[geo_pos] = dimension[geo_dimension]["category"]["index"][geo]
                coords[time_pos] = dimension[time_dimension]["category"]["index"][time]
                if value_pos is not None:
                    coords[value_pos] = dimension[value_dimension]["category"]["index"][value_key]
                index = 0
                for size, coord in zip(sizes, coords):
                    index = index * size + coord
                value = values[index]
                row = {geo_dimension: geo, time_dimension: time, "value": value}
                if value_pos is not None and value_dimension is not None:
                    row[value_dimension] = value_key
                rows.append(row)
    return rows


def infer_column(headers: Iterable[str], candidates: Iterable[str]) -> str | None:
    lower_headers = {header.lower(): header for header in headers}
    for candidate in candidates:
        if candidate.lower() in lower_headers:
            return lower_headers[candidate.lower()]
    for header in headers:
        header_lower = header.lower()
        for candidate in candidates:
            if candidate.lower() in header_lower:
                return header
    return None


def parse_reference_date_flexible(value: str) -> date:
    try:
        return parse_reference_date(str(value))
    except ValueError:
        match = re.search(r"(19|20)\\d{2}", str(value))
        if match:
            return date(int(match.group(0)), 6, 30)
    raise ValueError(f"Unsupported time value: {value}")


def map_rows(
    rows: Iterable[dict],
    geo_type: str,
    dataset_id: str,
    geo_code_column: str,
    time_column: str,
    value_column: str,
    measure_column: str | None,
) -> list[dict]:
    mapped = []
    for row in rows:
        reference_date = parse_reference_date_flexible(row[time_column])
        mapped.append(
            {
                "geo_type": normalize_geo_type(geo_type),
                "geo_code": str(row[geo_code_column]).strip(),
                "reference_date": reference_date.isoformat(),
                "population_value": int(float(row[value_column])),
                "dataset_id": dataset_id,
                "measure": row.get(measure_column) if measure_column else None,
            }
        )
    return mapped
